File: 프로그래머스/2/131127. 할인 행사/할인 행사.cpp
#include <iostream>
#include <algorithm>
#include <string>
#include <vector>

using namespace std;

bool isFindAll(vector<int> number) {
    for (auto elem : number) {
        if (elem != 0) return false;
    }
    return true;
}

int solution(vector<string> want, vector<int> number, vector<string> discount) {
    int answer = 0;
    vector<string> origin_want = want;
    vector<int> origin_number = number;
    
    for(int start = 0; start < discount.size() - 9; start++) {
        want = origin_want;
        number = origin_number;
        for(int i = start; i < start + 10; i++) {
            auto it = find(want.begin(), want.end(), discount[i]);
            if (it != want.end()) {
                number[it - want.begin()]--;
                if (number[it - want.begin()] == 0) {
                    want[it - want.begin()] = "";
                }
            }
        }

        if (isFindAll(number)) answer++;

    }
    
    return answer;
}

File: 프로그래머스/2/131127. 할인 행사/할인 행사.py
import copy

def solution(want, number, discount):
    answer = 0
    # 구매 품목 딕셔너리 생성
    dic = {}
    for goods, num in zip(want, number):
        dic[goods] = num
        
    start_idx = 0
    while start_idx + 10 <= len(discount):
        dic_copy = dic.copy()
        for idx in range(start_idx, start_idx + 10):
            item = discount[idx]
            if item in dic_copy and dic_copy[item] != 0:
                dic_copy[item] -= 1
                
        # 다 살 수 있는지 확인
        for k, v in dic_copy.items():
            if v != 0:
                break
        else:
            answer += 1
            
        start_idx += 1
        
    return answer